v1.3 80 Price histogram

trxe opened the pull request:
closes #80.

Summary:
- Added new command `StatCommand` and its parser
- Refactored `CommandResult` to contain an `Optional<UiElement>` field
- `HistogramStat` is a `UiElement` which creates a `JFreeChart` histogram on reaching the `StatWindow` handler.
- `MainWindow` modified as necessary.

guangxuan commented on src/main/java/seedu/address/logic/commands/StatCommand.java:
```diff
@@ -0,0 +1,71 @@
+        } else if (!showBuyer & showProperty) {
+            msgArgs = PROPERTY_WORD;
+        }
+        Stat stat = new HistogramStat(buyerList, propertyList, showBuyer, showProperty, msgArgs);
```
we can do some empty list checking here. Throw some exception to be caught by LogicManager.execute

guangxuan commented on src/main/java/seedu/address/logic/parser/StatCommandParser.java:
```diff
@@ -0,0 +1,42 @@
+            ParserUtil.assertPreambleArgsCount(userInput, NUM_OF_PREAMBLE_ARGS);
+            actor = ParserUtil.parseActor(userInput, ACTOR_POSITIONAL_INDEX);
+        } catch (ParseException pe) {
+            if (userInput.equals(new String())) {
```
Slight nitpicking here but we could use .isEmpty()

guangxuan commented on src/main/java/seedu/address/model/stats/HistogramStat.java:
```diff
@@ -0,0 +1,60 @@
+
+        HistogramDataset dataset = new HistogramDataset();
+        if (showBuyer) {
+            dataset.addSeries("buyers", buyerPrices, BIN_COUNT);
```
empty lists seem to cause exceptions here. 
Interestingly `stat` is not NULL within MainWindow but it just cannot draw here.
Exception in thread "JavaFX Application Thread" java.lang.IllegalArgumentException: Null or zero length 'values' argument.
	at org.jfree.data.statistics.HistogramDataset.getMinimum(HistogramDataset.java:198)
	at org.jfree.data.statistics.HistogramDataset.addSeries(HistogramDataset.java:113)
	at seedu.address.model.stats.HistogramStat.create(HistogramStat.java:48)
	at seedu.address.ui.StatWindow.show(StatWindow.java:60)
	at seedu.address.ui.MainWindow.handleStat(MainWindow.java:173)

trxe replied:
Alright sure, handling empty lists as explained in the comment above.

guangxuan commented on src/main/java/seedu/address/ui/UiElement.java:
```diff
@@ -0,0 +1,5 @@
+package seedu.address.ui;
+
```
Could add a javadoc here. Short one will do.

guangxuan commented on src/main/java/seedu/address/model/stats/Stat.java:
```diff
@@ -0,0 +1,11 @@
+import org.jfree.chart.JFreeChart;
+
+import seedu.address.ui.UiElement;
+
```
Could add a javadoc here. Short one will do.

guangxuan commented on src/main/java/seedu/address/model/stats/HistogramStat.java:
```diff
@@ -0,0 +1,60 @@
+import javafx.collections.ObservableList;
+import seedu.address.model.property.Buyer;
+import seedu.address.model.property.Property;
+
```
Could add a javadoc here.

guangxuan commented on src/main/java/seedu/address/logic/commands/StatCommand.java:
```diff
@@ -0,0 +1,81 @@
+import seedu.address.model.property.Property;
+import seedu.address.model.stats.HistogramStat;
+import seedu.address.model.stats.Stat;
+
```
oh I forgot about this javadocs too

trxe commented:
New updates:
- Added Priceable interface to Buyer and Property for easier handling in statistics.
- Fixed visual bugs for histogram by implementing creation and labelling of bins.

guangxuan commented on src/main/java/seedu/address/model/stats/HistogramStat.java:
```diff
@@ -0,0 +1,139 @@
+     * @return price histogram.
+     */
+    public JFreeChart create() {
+        assert !buyerList.isEmpty() && !propertyList.isEmpty();
```
btw this assertion fails when we try to do `stat buyer` with no `properties`
or `stat` with no `properties`

guangxuan reviewed (changes requested):
pretty close already, this should be the last bug
